ext/ScreenShot/ScreenShot.h

```diff
@@ -5,72 +5,186 @@
+					auto gpuImageView = driver->getGPUObjectsFromAssets(&imageView.get(), &imageView.get() + 1)->front();
+
+					return std::move(gpuImageView);
+				};
+
+				auto gpuImageViewDepthBuffer = createAttachement(false);
+				auto gpuImageViewColorBuffer = createAttachement(true);
+
+				auto frameBuffer = driver->addFrameBuffer();
+				frameBuffer->attach(video::EFAP_DEPTH_ATTACHMENT, std::move(gpuImageViewDepthBuffer));
+				frameBuffer->attach(video::EFAP_COLOR_ATTACHMENT0, std::move(gpuImageViewColorBuffer));
+
+				return frameBuffer;
+			};
+
+			/*
+				Download mip level image with gpu image usage and save it to IGPUBuffer.
+				Because of the fence placed by driver the function stalls the CPU 
+				to wait on the GPU to finish, beware of that.
+				@see video::IDriverFence
+			*/
+
+			//! TODO: HANDLE UNPACK ALIGNMENT
+			core::smart_refctd_ptr<video::IDriverFence> downloadImageMipLevel(video::IDriver* driver, video::IGPUImage* source, video::IGPUBuffer* destination, uint32_t sourceMipLevel = 0u, size_t destOffset = 0ull, bool implicitflush = true)
+			{
+				// will change this, https://github.com/buildaworldnet/IrrlichtBAW/issues/148
+				if (isBlockCompressionFormat(source->getCreationParameters().format))
+					return nullptr;
+
+				auto extent = source->getMipSize(sourceMipLevel);
+				video::IGPUImage::SBufferCopy pRegions[1u] = { {destOffset,extent.x,extent.y,{static_cast<asset::IImage::E_ASPECT_FLAGS>(0u),sourceMipLevel,0u,1u},{0u,0u,0u},{extent.x,extent.y,extent.z}} };
+				driver->copyImageToBuffer(source, destination, 1u, pRegions);
+
+				return driver->placeFence(implicitflush);
+			}
+
+			/*
+				Create a ScreenShot with gpu image usage and save it to a file.
+			*/
+
+			bool createScreenShot(core::smart_refctd_ptr<IrrlichtDevice> device, const video::IGPUImageView* gpuImageView, const std::string& outFileName)
+			{
+				auto driver = device->getVideoDriver();
+				auto assetManager = device->getAssetManager();
+
+				auto fetchedImageViewParmas = gpuImageView->getCreationParameters();
+				auto gpuImage = fetchedImageViewParmas.image;
+				auto fetchedImageParams = gpuImage->getCreationParameters();
+				auto image = asset::ICPUImage::create(std::move(fetchedImageParams));
+
+				auto trueBufferRowLength = asset::IImageAssetHandlerBase::calcPitchInBlocks(fetchedImageParams.extent.width, asset::getTexelOrBlockBytesize(fetchedImageParams.format));
+
+				auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<asset::ICPUImage::SBufferCopy>>(1u);
+				asset::ICPUImage::SBufferCopy& region = regions->front();
+
+				region.imageSubresource.mipLevel = 0u;
+				region.imageSubresource.baseArrayLayer = 0u;
+				region.imageSubresource.layerCount = 1u;
+				region.bufferOffset = 0u;
+				region.bufferRowLength = trueBufferRowLength;
+				region.bufferImageHeight = 0u;
+				region.imageOffset = { 0u, 0u, 0u };
+				region.imageExtent = image->getCreationParameters().extent;
```

Comment:
if you dont plan on supporting downloading multi-mip level image views then assert that they only have one mipmap level.... or make more regions

also you're not specifying the parameters correctly given the image view, the image view defines the subresource (layer offset, miplevel offset, and their counts), if I pass a really large (layers and mips) gpu image to this function I will end up having an ICPUImage with just as many layers and mips, except that it will only be partially covered by one region

Reply:
I specify image view params bellow

Reply:
maybe add a note in docs, that it will handle 0 miplevel only?

Reply:
> maybe add a note in docs, that it will handle 0 miplevel only?

yes and assert any assumptions you make about the input